### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by YarinsHyper**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,3 @@
-# api-gateway
\ No newline at end of file
+# api-gateway
+this is the api-gateway that's connected to birthday-service through grpc.
+made by yarin.
```
Comment (Shahar-Y) [inline README.md]: Add a link to the birthday-service so it is clear

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,11 @@
+module github.com/yarinBenisty/api-gateway
+
+go 1.16
+
+require (
+	github.com/gin-gonic/gin v1.7.2
+	github.com/google/uuid v1.2.0
+	github.com/gorilla/mux v1.8.0 // indirect
```
Comment (Shahar-Y) [inline go.mod]: run `go mod tidy` for a cleaner code

File: main.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+//functions in the bd-service proto file
+type Router struct {
+	client bpb.BirthdayFunctionsClient
+}
+
+//function that creates a new client and connects
+//to the bd-service client through the proto file
+func initClientConnection() bpb.BirthdayFunctionsClient {
+	conn, err := grpc.Dial(
+		"localhost:8000",
```
Comment (Shahar-Y) [inline main.go]: this will not always be localhost, and the port should be an environment variable

File: main.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	r := &Router{}
+	r.client = initClientConnection()
+
+	mainRouter := gin.Default()
+	mainRouter.POST("/api/createBirthday", r.CreateBirthday)
+	mainRouter.GET("/api/getBirthday/query", r.GetBirthday)
+	mainRouter.POST("/api/updateBirthday", r.UpdateBirthday)
+	mainRouter.DELETE("/api/deleteBirthday/query", r.DeleteBirthday)
+
+	err := mainRouter.Run(":9000")
```
Comment (Shahar-Y) [inline main.go]: port should be an environment variable

File: main.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+//function that deletes a certain birthday object by personal number
+func (r *Router) DeleteBirthday(c *gin.Context) {
+	personalNumber := c.Query("personalNumber")
+
+	request := &bpb.DeleteBirthdayRequest{PersonalNumber: personalNumber}
+	res, err := r.client.DeleteBirthday(c, request)
+	if err != nil {
+		fmt.Println("delete birthday method failed. error: ", err)
```
Comment (Shahar-Y) [inline main.go]: so after the error is printed, the result is success? (200)
you should return an error message and an error code to display to the client

File: main.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	c.JSON(200, res)
+}
+
+//function that updates a birthday object by personalNumber
+func (r *Router) UpdateBirthday(c *gin.Context) {
+	personalNumber := c.Request.FormValue("personalNumber")
+	name := c.Request.FormValue("name")
+	date := c.Request.FormValue("date")
```
Comment (Shahar-Y) [inline main.go]: all of these words should be constants in the beginning of the file

File: main.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	client := bpb.NewBirthdayFunctionsClient(conn)
+
+	return client
+}
+
+//function that inserts a birthday object
+func (r *Router) CreateBirthday(c *gin.Context) {
+	name := c.Request.FormValue("name")
+	personalNumber := c.Request.FormValue("personalNumber")
+	date := c.Request.FormValue("date")
```
Comment (Shahar-Y) [inline main.go]: these should be constants

File: main.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	c.JSON(200, res)
+}
+
+func main() {
+	r := &Router{}
+	r.client = initClientConnection()
+
+	mainRouter := gin.Default()
+	mainRouter.POST("/api/createBirthday", r.CreateBirthday)
+	mainRouter.GET("/api/getBirthday/query", r.GetBirthday)
```
Comment (Shahar-Y) [inline main.go]: no need for `query`

File: main.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	mainRouter.POST("/api/updateBirthday", r.UpdateBirthday)
+	mainRouter.DELETE("/api/deleteBirthday/query", r.DeleteBirthday)
```
Comment (Shahar-Y) [inline main.go]: no need for `query`

File: main.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+//function that read/get all birthday objects
+func (r *Router) GetBirthday(c *gin.Context) {
+	personalNumber := c.Query("personalNumber")
+
+	request := &bpb.GetBirthdayRequest{PersonalNumber: personalNumber}
+	res, err := r.client.GetBirthday(c, request)
+	if err != nil {
+		fmt.Println("get birthday method failed. error: ", err)
```
Comment (Shahar-Y) [inline main.go]: Other than printing, how is the error handled? Do you return an error to the client?

File: main.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+//function that updates a birthday object by personalNumber
+func (r *Router) UpdateBirthday(c *gin.Context) {
+	personalNumber := c.Request.FormValue("personalNumber")
+	name := c.Request.FormValue("name")
+	date := c.Request.FormValue("date")
+
+	request := &bpb.UpdateBirthdayRequest{PersonalNumber: personalNumber, Name: name, Date: date}
+	res, err := r.client.UpdateBirthday(c, request)
+	if err != nil {
+		fmt.Println("update birthday method failed. error: ", err)
```
Comment (Shahar-Y) [inline main.go]: Error handling

File: main.go [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+func (r *Router) CreateBirthday(c *gin.Context) {
+	name := c.Request.FormValue("name")
+	personalNumber := c.Request.FormValue("personalNumber")
+	date := c.Request.FormValue("date")
+
+	request := &bpb.CreateBirthdayRequest{
+		PersonalNumber: personalNumber, Name: name, Date: date}
+	res, err := r.client.CreateBirthday(c, request)
+	if err != nil {
+		fmt.Println("create birthday method failed. error: ", err)
```
Comment (Shahar-Y) [inline main.go]: Error handling

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
I see you do not have the linter enabled. 
In vscode go to File>Preferences>Settings>Go. under User tab, change the linting tool to golint.
After that, correct the problems.

Good progress, needs some patching :+1:

File: README.md (hunk 1 above)
Comment (YarinsHyper) [reply README.md]: done

File: go.mod (hunk 2 above)
Comment (YarinsHyper) [reply go.mod]: done

File: main.go (hunk 3 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 4 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 5 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 6 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 7 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 10 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 11 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 12 above)
Comment (YarinsHyper) [reply main.go]: done

**Pull request merged**